src/Slacklinker/Sender/UserDataUpload.hs

```diff
@@ -0,0 +1,108 @@
+module Slacklinker.Sender.UserDataUpload (doUploadUserData) where
+
+import Control.Monad.Catch (MonadThrow (..))
+import Data.Aeson (eitherDecode)
+import Database.Persist
+import Network.URI (parseURI)
+import OpenTelemetry.Instrumentation.HttpClient (
+  Request (..),
+  httpClientInstrumentationConfig,
+  httpLbs,
+  requestFromURI,
+  responseBody,
+  setRequestCheckStatus,
+ )
+import Slacklinker.App (App (..), HasApp (..), appSlackConfig, runDB)
+import Slacklinker.Import
+import Slacklinker.Models
+import Slacklinker.Sender.Internal
+import Slacklinker.Sender.Types
+import Slacklinker.Sender.UserDataUpload.Types
+import Slacklinker.Settings (SlacklinkerSetting (..), SlacklinkerSettingTag (..), getSetting)
+import Slacklinker.Types (SlackToken (..))
+import Web.Slack (usersListAll)
+import Web.Slack.Conversation
+import Web.Slack.Files.Types (FileMode (..), FileObject (..), FileObjectVisible (..))
+import Web.Slack.Pager (loadingPage)
+import Web.Slack.User
+import Web.Slack.User qualified as User
+
+updateSlackUsersList :: (HasApp m, MonadIO m) => WorkspaceMeta -> m ()
+updateSlackUsersList wsInfo = do
+  slackConfig <- appSlackConfig wsInfo.token
+  users_ <- liftIO $ usersListAll slackConfig (def {User.listReqLimit = Just 100})
+  users <- liftIO $ loadingPage users_ convertPage
+
+  runDB $ for_ users \user ->
+    let email = user.userProfile >>= (.profileEmail)
+     in upsertBy
+          (UniqueKnownUser wsInfo.workspaceId user.userId)
+          KnownUser
+            { workspaceId = wsInfo.workspaceId
+            , slackUserId = user.userId
+            , email
+            , emoji = Nothing
+            , githubUsername = Nothing
+            }
+          [KnownUserEmail =. email]
+
+  pure ()
```

Comment:
Looks like a redundant `pure ()`, since you're not discarding the result of the `runDB` call above.